Gateway returns an HTTP status code 500 status when query parameters are not properly encoded

java.lang.IllegalArgumentException: Illegal character in query at index 49: http://localhost:8082/test/my_team?q=myparam:test AND myotherparam:12

	at java.net.URI.create(URI.java:852)
	at org.apache.http.client.fluent.Request.Get(Request.java:86)
	at io.gravitee.gateway.standalone.QueryParametersTest.call_get_query_params_spaces(QueryParametersTest.java:75)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at io.gravitee.gateway.standalone.junit.stmt.ApiDeployerStatement.evaluate(ApiDeployerStatement.java:63)
	at io.gravitee.gateway.standalone.junit.stmt.ApiPublisherStatement.evaluate(ApiPublisherStatement.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)

Caused by: java.net.URISyntaxException: Illegal character in query at index 49: http://localhost:8082/test/my_team?q=myparam:test AND myotherparam:12
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3111)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at java.net.URI.create(URI.java:850)
	... 31 more

--- a/gravitee-gateway-http-client/gravitee-gateway-http-client-core/src/main/java/io/gravitee/gateway/http/core/invoker/AbstractHttpInvoker.java
+++ b/gravitee-gateway-http-client/gravitee-gateway-http-client-core/src/main/java/io/gravitee/gateway/http/core/invoker/AbstractHttpInvoker.java
@@ -27,7 +27,9 @@ import io.gravitee.gateway.http.core.logger.LoggableClientRequest;
 import io.gravitee.gateway.http.core.logger.LoggableClientResponse;
 import org.springframework.beans.factory.annotation.Autowired;
 
+import java.io.UnsupportedEncodingException;
 import java.net.URI;
+import java.net.URLEncoder;
 import java.util.Map;
 import java.util.stream.Collectors;
 
@@ -114,19 +116,24 @@ public abstract class AbstractHttpInvoker implements Invoker {
                         .insert(0, endpointUri.toString());
 
         if (request.parameters() != null && ! request.parameters().isEmpty()) {
-            requestURI.append('?');
+            StringBuilder query = new StringBuilder();
+            query.append('?');
 
             for(Map.Entry<String, String> queryParam : request.parameters().entrySet()) {
-                requestURI.append(queryParam.getKey());
+                query.append(queryParam.getKey());
                 if (queryParam.getValue() != null && !queryParam.getValue().isEmpty()) {
-                    requestURI.append('=').append(queryParam.getValue());
+                    query.append('=').append(queryParam.getValue());
                 }
 
-                requestURI.append('&');
+                query.append('&');
             }
 
-            // Removing latest & separator
-            requestURI.deleteCharAt(requestURI.length() - 1);
+            // Removing latest & separator and encode query parameters
+            try {
+                requestURI.append(URLEncoder.encode(query.deleteCharAt(query.length() - 1).toString(), "UTF-8"));
+            } catch (UnsupportedEncodingException e) {
+                e.printStackTrace();
+            }
         }
 
         return URI.create(requestURI.toString());

--- a/gravitee-gateway-standalone/gravitee-gateway-standalone-container/src/test/java/io/gravitee/gateway/standalone/QueryParametersTest.java
+++ b/gravitee-gateway-standalone/gravitee-gateway-standalone-container/src/test/java/io/gravitee/gateway/standalone/QueryParametersTest.java
@@ -25,6 +25,9 @@ import org.apache.http.client.fluent.Request;
 import org.apache.http.client.fluent.Response;
 import org.junit.Test;
 
+import java.net.URLEncoder;
+import java.nio.charset.Charset;
+
 import static org.junit.Assert.assertEquals;
 
 /**
@@ -55,6 +58,21 @@ public class QueryParametersTest extends AbstractGatewayTest {
     public void call_get_query_params_emptyvalue() throws Exception {
         String query = "query";
         Request request = Request.Get("http://localhost:8082/test/my_team?" + query);
+        Response response = request.execute();
+
+        HttpResponse returnResponse = response.returnResponse();
+        assertEquals(HttpStatus.SC_OK, returnResponse.getStatusLine().getStatusCode());
+
+        String responseContent = StringUtils.copy(returnResponse.getEntity().getContent());
+        assertEquals(query, responseContent);
+    }
+
+    @Test
+    public void call_get_query_params_spaces() throws Exception {
+        String query = "q=myparam:test AND myotherparam:12";
+        String encodedQuery = query; //URLEncoder.encode(query, "UTF-8");
+
+        Request request = Request.Get("http://localhost:8082/test/my_team?" + encodedQuery);
         Response response = request.execute();
 
         HttpResponse returnResponse = response.returnResponse();

